ouroboros-consensus-test/bench-storage/Bench/Ouroboros/Consensus/Storage/LedgerDB/HD.hs

```diff
@@ -0,0 +1,631 @@
+    pure $ Model mempty 0 (TT.valuesFromList kvs) nrInitialValues 0
+
+genKeyValue :: Arbitrary v => k -> Gen (k, v)
+genKeyValue x = (x,) <$> arbitrary
+
+{-------------------------------------------------------------------------------
+  Command generator: Main generators
+-------------------------------------------------------------------------------}
+
+-- | A stateful generator for sequences of commands.
+--
+-- Note: We return the @'Model'@ because we might want to check invariants on
+-- it.
+generator :: (Eq v, Arbitrary v) => GenConfig -> Gen ([Cmd 'New Int v], Model Int v)
+generator conf = do
+  m <- genInitialModel conf
+  runStateT (genCmds conf) m
+
+-- | Like @'generator'@, but the order of operations is deterministic.
+generator' :: (Eq v, Arbitrary v) => GenConfig -> Gen ([Cmd 'New Int v], Model Int v)
+generator' conf = do
+  m <- genInitialModel conf
+  runStateT (genCmds' conf) m
+
+-- | Adapt a @GenConfig@ured generator to multiply the configured @nrCommands@
+-- parameter with a factor that depends on the QuickCheck size parameter.
+makeSized ::
+     (GenConfig -> Gen a)
+  -> GenConfig
+  -> Gen a
+makeSized fgen conf = do
+  n <- getSize
+  k <- choose (0, n)
+  let
+    p :: Float
+    p =
+      if n == 0 then
+        0
+      else
+        fromIntegral k / fromIntegral n
+    nr :: Float
+    nr = fromIntegral (nrCommands conf)
+
+  fgen (conf {nrCommands = round $ p * nr})
+
+type CmdGen v a = StateT (Model Int v) Gen a
+
+-- | Generate a list of commands in non-deterministic order.
+genCmds :: (Eq v, Arbitrary v) => GenConfig -> CmdGen v [Cmd 'New Int v]
+genCmds conf = mapM (const $ genCmd conf) [1 .. nrCommands conf]
+
+-- | Generate a list of commands in deterministic order.
+genCmds' :: forall v. (Eq v, Arbitrary v) => GenConfig -> CmdGen v [Cmd 'New Int v]
+genCmds' conf = do
+    cmds <- go (nrCommands conf)
+    pure $ Exn.assert (length cmds == nrCommands conf) cmds
+  where
+    go n = do
+      xs <- genN 10 (genForwardPushFlush (genN 10 genForwardPush))
+      x  <- guard (singleton <$> genRollback conf)
+
+      let
+        xs0 = xs ++ x
+        len = length xs0
+
+      if length xs0 >= n then
+        pure $ take n xs0
+      else do
+        xs0' <- go (n - len)
+        pure $ xs0 ++ xs0'
+
+    genForwardPushFlush :: CmdGen v [Cmd 'New Int v] -> CmdGen v [Cmd 'New Int v]
+    genForwardPushFlush gen = do
+      fps <- gen
+      f   <- guard (singleton <$> genFlush conf)
+
+      pure $ fps ++ f
+
+    genForwardPush :: CmdGen v [Cmd 'New Int v]
+    genForwardPush = do
+      f <- guard (singleton <$> genForward conf)
+      p <- guard (singleton <$> genPush conf)
+      pure $ f ++ p
+
+    guard :: CmdGen v [Cmd 'New Int v] -> CmdGen v [Cmd 'New Int v]
+    guard gen = do
+      nrg <- gets nrGenerated
+      let nrc = nrCommands conf
+
+      if nrg < nrc then
+        gen
+      else
+        pure []
+
+    singleton :: a -> [a]
+    singleton x = [x]
```

Comment:
A newer version of `base` should have a `singleton` function. Add a FIXME to remove this function when we depend on this newer version of `base`.